core/src/main/java/de/danoeh/antennapod/core/service/FeedUpdateWorker.java

```diff
@@ -31,6 +56,26 @@ public Result doWork() {
         ClientConfig.initialize(getApplicationContext());
 
         if (NetworkUtils.networkAvailable() && NetworkUtils.isFeedRefreshAllowed()) {
+            List<Feed> feeds = new ArrayList<>();
+            Map<Long, FeedItem> lastItemsMap = new HashMap<>();
+
+            for (Feed feed : DBReader.getFeedList()) {
+                FeedPreferences prefs = feed.getPreferences();
+                if (prefs.getKeepUpdated() && prefs.getShowNotification()) {
+                    List<FeedItem> outdatedFeedItems = DBReader.getFeedItemList(feed);
```

Comment:
In the worst case, this will load all episodes of all feeds. That will lead to out-of-memory errors when using a big list of feeds. Something similar was done in another location and I just recently found out that this is responsible for a pretty decent number of crashes. So there must be some way to do that without storing all episodes.

Do we actually need to store what episodes are new? I think the "new" flag in the database is basically meant to be used for exactly that purpose: an episode that users have not interacted with yet. Using the "new" flag will also fix this case:

> Currently if there a notification for 2 episodes and another is found, the notification gets updated with only the newly found episodes (e.g 1) instead of adding the previous episodes

Reply:
I thought about that, but basically you would get a notification you would end up with a lot of notifications, because you would get a notification each time there is a background update and there is some episode marked as new. So you gotta go through your podcasts and manually remove all the new flags.

Reply:
How about using the database to store the most recent episode ID for each subscription (or alternatively, maybe more efficiently, simply the number of episodes)? If the most recent episode changes during refresh, display a notification of all "new" episodes. That way, we do not need to store another flag per episode (which needs more space when multiplying with 20k+ items), only one per subscription. Also, this fixes the case that the notification is not updated properly when there are additional new items.

It still might display a notification for a new item twice but only if there is an additional new item. I think this is pretty reasonable and also is consistent with the idea of "new" episodes, being the ones you have not interacted with yet.

Reply:
Okay, so I just had a look at what you recently changed. I like that it works without new database entries that store the episode state. How about making the `NewEpisodesNotification` class a normal object that is created when the download service starts and then shows the notification when the download service is stopped? I think static variables that store actual user data should be avoided when possible because it makes it harder to add unit tests.

Reply:
While implementing this I also had the idea of using the new Flag but clearing it every-time the feed gets opened or notification gets dismissed and giving the option to clear it via an notification action. I didn't do this because this would change the behavior of the new flag and to me it feels like issue which should be solved after its own discussion in its own PR.

Reply:
My idea was not to clear the new flag. Just notify for all "new" flagged episodes whenever an additional episode is released. This fixes this problem:

> Currently if there a notification for 2 episodes and another is found, the notification gets updated with only the newly found episodes (e.g 1) instead of adding the previous episodes

I think that it is okay to display all "new" flagged ones again (even if they might be displayed twice, then) because they are only displayed twice if an additional one was released. I also think that it is okay to display them again - being flagged as "new" means that the user has not interacted with them yet.